File: app/src/main/java/test/alexzander/swipetodelete/ContactAdapterUtils.java
package test.alexzander.swipetodelete;

import android.animation.Animator;
import android.animation.ValueAnimator;
import android.view.View;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by AlexZandR on 4/24/17
 */

public abstract class ContactAdapterUtils {

    public static final long DELETING_DURATION = 3000;
    public static final long PENDING_DURATION = DELETING_DURATION - 150;
    public static final float DEVICE_SCREEN_WIDTH = 1080F;

    private ContactAdapterUtils() {
    }

    public static List<ItemContact> prepareContactList(int count) {
        List<ItemContact> result = new ArrayList<>(count);
        for (int i = 0; i < count; i++) {
            result.add(new ItemContact("User Name " + String.valueOf(i), "+1234567" + String.valueOf(i)));
        }
        return result;
    }

    public static ValueAnimator createAnimator(final View view, final ItemContact contact) {
        return initAnimator(view, contact, null);
    }

    public static ValueAnimator initAnimator(final View view, final ItemContact contact,
                                             ValueAnimator animator) {
        if (animator == null) {
            animator = ValueAnimator.ofFloat(contact.posX, DEVICE_SCREEN_WIDTH * contact.direction);
        } else {
            animator.removeAllUpdateListeners();
            animator.removeAllListeners();
            animator.setFloatValues(contact.posX, DEVICE_SCREEN_WIDTH * contact.direction);
        }
        animator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                final float posX = (Float) animation.getAnimatedValue();
                view.setX(posX);
                contact.posX = posX;
            }
        });
        animator.addListener(new Animator.AnimatorListener() {
            @Override
            public void onAnimationStart(Animator animation) {
                contact.isRunningAnimation = true;
                view.setVisibility(View.VISIBLE);
            }

            @Override
            public void onAnimationEnd(Animator animation) {
                contact.isRunningAnimation = false;
            }

            @Override
            public void onAnimationCancel(Animator animation) {
                clearContact(contact);
                clearView(view);
            }

            @Override
            public void onAnimationRepeat(Animator animation) {
            }
        });
        animator.setDuration((long) (DELETING_DURATION * (DEVICE_SCREEN_WIDTH - contact.posX * contact.direction) / DEVICE_SCREEN_WIDTH));
        return animator;
    }

    public static void clearAnimator(final ValueAnimator animator) {
        if (animator != null) {
            animator.cancel();
            animator.removeAllUpdateListeners();
            animator.removeAllListeners();
        }
    }

    public static void clearContact(final ItemContact contact) {
        if (contact != null) {
            contact.isPendingDelete = false;
            contact.isRunningAnimation = false;
            contact.posX = 0;
        }
    }

    public static void clearView(final View view) {
        if (view != null) {
            view.setX(0);
            view.setVisibility(View.GONE);
        }
    }
}

File: app/src/main/java/test/alexzander/swipetodelete/ContactAdapterUtils.kt
package test.alexzander.swipetodelete

import android.animation.Animator
import android.animation.ValueAnimator
import android.view.View

import java.util.ArrayList

/**
 * Created by AlexZandR on 4/24/17
 */

object ContactAdapterUtils {

    val DELETING_DURATION: Long = 3000
    val PENDING_DURATION = DELETING_DURATION - 150

    fun prepareContactList(count: Int): MutableList<ItemContact> {
        val result = ArrayList<ItemContact>(count)
        (0..count - 1).mapTo(result) { ItemContact("User Name " + it.toString(), "+1234567" + it.toString()) }
        return result
    }

    fun createAnimator(view: View, contact: ItemContact): ValueAnimator {
        return initAnimator(view, contact, null)
    }

    fun initAnimator(view: View, contact: ItemContact,
                     animator: ValueAnimator?): ValueAnimator {
        var animator = animator
        val screenWidth = MainActivity.sDeviceScreenWidth.toFloat()
        if (animator == null) {
            animator = ValueAnimator.ofFloat(contact.posX, screenWidth * contact.direction!!)
        } else {
            animator.removeAllUpdateListeners()
            animator.removeAllListeners()
            animator.setFloatValues(contact.posX, screenWidth * contact.direction!!)
        }
        animator!!.addUpdateListener { animation ->
            val posX = animation.animatedValue as Float
            view.x = posX
            contact.posX = posX
        }
        animator.addListener(object : Animator.AnimatorListener {
            override fun onAnimationStart(animation: Animator) {
                contact.isRunningAnimation = true
                view.visibility = View.VISIBLE
            }

            override fun onAnimationEnd(animation: Animator) {
                contact.isRunningAnimation = false
            }

            override fun onAnimationCancel(animation: Animator) {
                clearContact(contact)
                clearView(view)
            }

            override fun onAnimationRepeat(animation: Animator) {}
        })
        animator.duration = (DELETING_DURATION * (screenWidth - contact.posX.toFloat() * contact.direction!!.toFloat()) / screenWidth).toLong()
        return animator
    }

    fun clearAnimator(animator: ValueAnimator?) {
        if (animator != null) {
            animator.cancel()
            animator.removeAllUpdateListeners()
            animator.removeAllListeners()
        }
    }

    fun clearContact(contact: ItemContact?) {
        if (contact != null) {
            contact.isPendingDelete = false
            contact.isRunningAnimation = false
            contact.posX = 0f
        }
    }

    fun clearView(view: View?) {
        if (view != null) {
            view.x = 0f
            view.visibility = View.GONE
        }
    }
}